### DISCUSSION TITLE
how to retrieve data from event

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#1032] [Q&A]
Hello,
I'm using a foreign object inside my glsp graph, it contains a select element. My issue here is that I couldn't retrieve data from `event.target.value` or(`options[index].value`). as it shows me `Property 'value' does not exist on type 'EventTarget'`
```ts
const htmlElement: any = (
       <select
          name={node.id}
          id={node.id}
          ref={node.id}       
        >
          {options.length === 0 ? (
            <option value="">empty</option>
          ) : (
            options.map((option: string, index: number) => (
              <option key={index} value={option}>
                {option}
              </option>
            ))
      )}
        </select>
    );
    setNamespace(htmlElement, "http://www.w3.org/1999/xhtml");
    on(htmlElement, 'change', (event) =>            
    {
      console.log('----------',node.id);
      console.log('-----------',event.target.options);
      
      this.handlClick(node.id , node.id )
    })
``` 
I tried to do a casting:
```
const target = event.target as HTMLSelectElement
      const selectedValue = target.value
      console.log('--------------',selectedValue);
```
but it shows me this: `root INFO -------------- [object Object]`
Can you help me with this, thanks in advance

**Comment by martin-fleck-at** [comment]
@Adem-Kri I'm not sure how exactly you integrate that in your example as part of a view, could you please provide a little bit more context on your element configuration and view implementation?

In any case, your best shot to see what it is you are actually getting is by using any of the browser developer tools so you can set a breakpoint and see the object more clearly. To improve the output of `[object Object]`, you could probably use `JSON.stringify(target.value)` which should give you more details on the object if it can be serialized (e.g., does not have cyclic dependencies).

**Reply by Adem-Kri** [reply]
@martin-fleck-at I'm using it as a `foreign object` inside a vnode, later I bind that vnode using `configureModelElement()` and use it in my view:
```tsx
injectable()
export class HtmlView implements IView {
  @inject(TYPES.IActionDispatcher)
  protected actionDispatcher: GLSPActionDispatcher;
.
.
.
render(...):VNode{
...
let vnode: any;
    vnode = (
      <g transform="scale(...) translate(...)">
        <g id="workflow-diagram_0_..." transform="translate(...)">
          <g id="workflow-diagram_0_...">
            <foreignObject
              requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"
              x="0"
              y="0"
              height="25"
              width="60"
            >
              {htmlElement}
            </foreignObject>
          </g>
        </g>
      </g>
    );
   
    return vnode;
```
Also, I tried your solution but it shows me this : `root INFO -------------- "[object Object]"`

If there's a better way to implement my work I'm open for any advice, thank you

**Reply by martin-fleck-at** [reply]
I'm not entirely sure how your rendering works, I did have some problems with the `foreignObject` integration within the VNode but I managed to get it to run on my side a little bit differently so I could check on the events. For me, the event is triggered and event.target is an instance of HTMLSelectElement, so I could simply access the value like that.

This is what is looks like for me:
```
import virtualizeString from 'sprotty/lib/lib/virtualize';

@injectable()
export class CustomShape implements IView {
    render(node: Readonly<SShapeElement>, context: RenderingContext, args?: IViewArgs): VNode | undefined {
        const options = ['A', 'B', 'C'];

        const select = document.createElement('select');
        select.name = node.id;
        select.id = node.id;

        for (const option of options) {
            const opt = document.createElement('option');
            opt.value = option;
            opt.text = option;
            select.add(opt);
        }

        const selectVNode = virtualizeString(select.outerHTML);
        on(selectVNode!, 'change', event => {
            console.log(node.id);
            if (event.target instanceof HTMLSelectElement) {
                console.log(event.target.value);
            }
        });
        const vnode = (
            <g>
                <foreignObject requiredFeatures='http://www.w3.org/TR/SVG11/feature#Extensibility' height={200} width={400} x={0} y={0}>
                    {selectVNode}
                </foreignObject>
            </g>
        );
        return vnode;
    }
}
```

Does your target look different?

**Reply by Adem-Kri** [reply]
@martin-fleck-at , thank you for being passionate, 

I tried implementing your code but the `return` keeps showing me this : 
```
Type 'Element' is missing the following properties from type 'VNode': sel, data, children, elm, textts(2739)
``` 
by the way to avoid the problem of integrating foreignObject within VNode I added `const selectVNode: any` 

return issue is what prevents me from implementing your solution.

**Reply by martin-fleck-at** [reply]
@Adem-Kri That is interesting, I tested this code on the current master of GLSP and I don't see the same error. Are you using an older version? How does your tsconfig look?

**Reply by Adem-Kri** [reply]
@martin-fleck-at here's how it looks like:
```json
{
  "compilerOptions": {
      "skipLibCheck": true,
      "declaration": true,
      "declarationMap": true,
      "noImplicitAny": true,
      "noEmitOnError": false,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "strictNullChecks": true,
      "experimentalDecorators": true,
      "emitDecoratorMetadata": true,
      "downlevelIteration": true,
      "resolveJsonModule": true,
      "module": "commonjs",
      "moduleResolution": "node",
      "target": "ES2017",
      "jsx": "react",
      "lib": [
          "ES2017",
          "dom"
      ],
      "sourceMap": true,
      "rootDir": "src",
      "outDir": "lib"
  },
  "include": ["src", "types"],
  "exclude": ["node_modules", "**/*.spec.ts"] 
}
```
do you advise me to upgrade or to do something, am working with glsp since a while but I'm still learning

**Reply by martin-fleck-at** [reply]
I don't see any major differences, our options are:
```
{
  "compilerOptions": {
    "declaration": true,
    "declarationMap": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "strictNullChecks": true,
    "noImplicitOverride": true,
    "noEmitOnError": false,
    "noUnusedLocals": true,
    "noImplicitReturns": true,
    "strict": true,
    "incremental": true,
    // Needs to be disabled to support property injection
    "strictPropertyInitialization": false,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "downlevelIteration": true,
    "module": "commonjs",
    "moduleResolution": "Node16",
    "target": "ES2017",
    "jsx": "react",
    "lib": ["ES2017", "dom"],
    "sourceMap": true,
    "types": ["node", "mocha", "reflect-metadata"],
    "rootDir": "src",
    "outDir": "lib",
    "baseUrl": ".",
    "reactNamespace": "JSX",
    "skipLibCheck": true,
    "resolveJsonModule": true
  },
  "include": ["src"]
}
```

Which version of GLSP are you using? I pushed my example to a branch, could you test whether this works on your end? https://github.com/eclipse-glsp/glsp-client/compare/master...select-view

**Reply by martin-fleck-at** [reply]
I believe someone else is seeing a similar error: https://github.com/eclipse-glsp/glsp/discussions/1033
